kumahq/kuma
# ci(stability): add target to focus and repeat new or modified tests

Added a new target to focus and repeat tests. I hope that this will help us preventing introducing flakes in the first place.

How to tests this change:
1. checkout this branch
2. create a new branch, add a tests (e.g. copy "should allow the traffic with default traffic permission" from universal tests and call it "should allow the traffic with default traffic permission 2", this is how I tested it)
3. run `make TEST_LIST_BRANCH_TO_DIFF_AGAINST=focus-and-repeat-tests test/e2e/prevent-flakes` #where `focus-and-repeat-tests` is the branch checked out in 1
4. the output should contain "This was attempt x of 10."

The next step would be to put this in `circleci` config and run alongside e2e tests, but for this to work the target needs to be in `master` (or `TEST_LIST_BRANCH_TO_DIFF_AGAINST`) branch.

In the future this could be extended to include tests that contents have changed (by using git diff, line number and matching that to the nearest tests) or even changed functionality by checking the dependency tree.

### Checklist prior to review



- [ ] [Link to relevant issue][1] as well as docs and UI issues --
- [ ] This will not break child repos: it doesn't hardcode values (.e.g "kumahq" as a image registry) and it will work on Windows, system specific functions like `syscall.Mkfifo` have equivalent implementation on the other OS --
- [ ] Tests (Unit test, E2E tests, manual test on universal and k8s) --
- [ ] Do you need to update [`UPGRADE.md`](../blob/master/UPGRADE.md)? --
- [ ] Does it need to be backported according to the [backporting policy](../blob/master/CONTRIBUTING.md#backporting)? --
- [ ] Do you need to explicitly set a [`> Changelog:` entry here](../blob/master/CONTRIBUTING.md#submitting-a-patch) or add a `ci/` label to run fewer/more tests?

[1]: https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword
